### PR TITLE
Basic resurrection of the mutation tester

### PR DESCRIPTION
**Pull request by lukechurch**
### Purpose

This does some basic cleanup of the test infrastructure. It is by no means perfect, but is in a better place for bug discovery after these changes land. Running these tests for a few seconds results in bugs being discovered.

The primary changes are:

- Removal of a false positive caused if there were no 
- Tighten the timings up so that more mutations can run
- First set of refactorings to localise functionality 

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] The level of testing this PR includes is appropriate

### Reviewers

(FILL ME IN) @pboyer

### CODE REVIEW

File: src/DynamoCoreWpf/TestInfrastructure/AbstractMutator.cs [hunk 1]
```diff
@@ -33,9 +34,37 @@ public virtual Type GetNodeType()
...
         public virtual int NumberOfLaunches
         {
-            get { return 1000; }
+            get { return 10; }
+        }
+
+
+
+
+        // TODO(lukechurch): Move this to a support class
```
Comment (pboyer) [inline src/DynamoCoreWpf/TestInfrastructure/AbstractMutator.cs]: Consider making YT tasks for all of the todo's. Curiously, tasks of this sort never seem to get done...

File: src/DynamoCoreWpf/TestInfrastructure/CodeBlockNodeMutator.cs [hunk 2]
```diff
@@ -61,7 +61,7 @@ public override bool RunTest(NodeModel node, EngineController engine, StreamWrit
 
                 DynamoViewModel.ExecuteCommand(runCancel);
             }));
-            Thread.Sleep(100);
+            Thread.Sleep(0);
```
Comment (pboyer) [inline src/DynamoCoreWpf/TestInfrastructure/CodeBlockNodeMutator.cs]: Can you explain why this `Thread.Sleep` call is present? There are different, guaranteed mechanisms for flushing the `Dispatcher`. For example, see `NodeViewCustomizationTests`.

File: src/DynamoCoreWpf/TestInfrastructure/AbstractMutator.cs [hunk 3]
```diff
@@ -33,9 +34,37 @@ public virtual Type GetNodeType()
...
+            SpinWaitForExecution();
+        }
+
+        // TODO(lukechurch): Move this to use event waiting rather than spin waiting
+        protected void SpinWaitForExecution()
+        {
+            while (!DynamoViewModel.HomeSpace.RunSettings.RunEnabled)
+            {
+                Thread.Sleep(10);
+            }
```
Comment (pboyer) [inline src/DynamoCoreWpf/TestInfrastructure/AbstractMutator.cs]: Ay, this feels brittle to me. What about something like this?

```csharp
var runDone = false;
Action a = () => runDone = true;
DynamoViewModel.HomeSpace.EvaluationComplete += a;
while (!runDone){
  Thread.Sleep(10);
}
DynamoViewModel.HomeSpace.EvaluationComplete -= a;
```

File: src/DynamoCoreWpf/TestInfrastructure/AbstractMutator.cs (hunk 1 above)
Comment (pboyer) [inline src/DynamoCoreWpf/TestInfrastructure/AbstractMutator.cs]: Please create YT tasks for todos.

**Comment by pboyer** [comment]
A few suggestions and then LGTM.

File: src/DynamoCoreWpf/TestInfrastructure/CodeBlockNodeMutator.cs (hunk 2 above)
Comment (lukechurch) [inline src/DynamoCoreWpf/TestInfrastructure/CodeBlockNodeMutator.cs]: These sleep calls will be repurposed to allow a best effort interleave tester. Not a great way of controlling the threads, but better than nothing.

I was reducing these both to improve the performance of the test and to try and avoid masking trivial race conditions.

File: src/DynamoCoreWpf/TestInfrastructure/AbstractMutator.cs (hunk 1 above)
Comment (lukechurch) [inline src/DynamoCoreWpf/TestInfrastructure/AbstractMutator.cs]: ack.

**Pull request merged**